### DISCUSSION TITLE
Green threads, channels, and concurrency in Nelua

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#209] [Q&A] [answered]
If I wanted to use concurrency and specifically [Communicating Sequential Processes](https://en.wikipedia.org/wiki/Communicating_sequential_processes), like Golang has as part of the language, how would I accomplish it with Nelua?

Any advice on this matter @edubart ?

**Comment by edubart** [comment, marked as answer]
No special advices for that, I guess you will need to come up with a library or framework with that goal. I would do concurrency in Nelua taking advantage of coroutines, because Nelua supports them. You can certainly do job/scheduler system using coroutines and multiple threads.

**Reply by stefanos82** [reply]
I can't find anything documented in https://nelua.io/overview/ ; do you have any demo to study it?

Update: Nevermind...I forgot to look in [Libraries#coroutine](https://nelua.io/libraries/#coroutine)

**Reply by stefanos82** [reply]
@edubart have you seen this project https://github.com/zhaojh329/lua-eco ?

It looks very interesting.

If I'm not mistaken, Nelua uses coroutines more or less the same as Lua.

Do you think this aforementioned project could be implemented in Nelua, but with the help of C libraries?

**Reply by edubart** [reply]
> Do you think this aforementioned project could be implemented in Nelua, but with the help of C libraries

Yes, I think you can implement event-loops with a scheduler on top of Nelua coroutines, even using multiple threads for maximum performance. But I would recommend experimenting with just a single thread using asynchronous operations first, keep things simple, it's easy to shoot yourself with multiple threads.

**Reply by stefanos82** [reply]
Fair point :+1:
